File: apps/electron/web/src/notice-markdown.ts
// 公告内容的 Markdown 子集解析：**粗体**、[文字](http(s) url)、行首 #~###### 标题。
// 其他 Markdown 标记原样保留为纯文本。解析前先剥离面板返回的 HTML 标签/实体。

export interface NoticeSpan {
  text: string
  bold: boolean
  href?: string
}

export type NoticeParagraph = NoticeSpan[]

export function parseNoticeMarkdown(content: string): NoticeParagraph[] {
  const text = stripNoticeHtml(content)
  if (!text) return []
  return text.split('\n').map(parseNoticeLine)
}

function stripNoticeHtml(value: string): string {
  return value
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/<\/(p|div|li|h[1-6])>/gi, '\n')
    .replace(/<[^>]+>/g, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
}

const HEADING_PATTERN = /^#{1,6}[ \t]+(.*)$/

function parseNoticeLine(line: string): NoticeParagraph {
  const heading = HEADING_PATTERN.exec(line)
  if (heading) return parseNoticeSpans(heading[1].trim(), true)
  return parseNoticeSpans(line, false)
}

function parseNoticeSpans(text: string, bold: boolean): NoticeSpan[] {
  const spans: NoticeSpan[] = []
  let plain = ''
  const flushPlain = () => {
    if (plain) {
      spans.push({ text: plain, bold })
      plain = ''
    }
  }
  let index = 0
  while (index < text.length) {
    if (!bold && text.startsWith('**', index)) {
      const end = text.indexOf('**', index + 2)
      if (end > index + 2) {
        flushPlain()
        spans.push(...parseNoticeSpans(text.slice(index + 2, end), true))
        index = end + 2
        continue
      }
    }
    if (text[index] === '[') {
      const link = matchNoticeLink(text, index)
      if (link) {
        flushPlain()
        spans.push({ text: link.text, bold, href: link.href })
        index = link.end
        continue
      }
    }
    plain += text[index]
    index += 1
  }
  flushPlain()
  return spans
}

interface NoticeLink {
  text: string
  href: string
  end: number
}

function matchNoticeLink(text: string, start: number): NoticeLink | null {
  const labelEnd = text.indexOf(']', start + 1)
  if (labelEnd < 0 || text[labelEnd + 1] !== '(') return null
  const hrefEnd = text.indexOf(')', labelEnd + 2)
  if (hrefEnd < 0) return null
  const label = text.slice(start + 1, labelEnd).trim()
  const href = cleanNoticeHref(text.slice(labelEnd + 2, hrefEnd))
  if (!label || !href) return null
  return { text: label, href, end: hrefEnd + 1 }
}

function cleanNoticeHref(raw: string): string {
  // 面板网页主题用 #eztheme-btn 锚点标记按钮样式，客户端渲染时去掉
  const href = raw.trim().replace(/#eztheme-btn$/i, '')
  if (!/^https?:\/\//i.test(href) || /\s/.test(href)) return ''
  return href
}

File: apps/android/src/main/kotlin/moe/telecom/xbclient/NoticeMarkdown.kt
package moe.telecom.xbclient

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.LinkAnnotation
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withLink
import androidx.compose.ui.text.withStyle
import top.yukonga.miuix.kmp.basic.Text

// 公告内容的 Markdown 子集解析：**粗体**、[文字](http(s) url)、行首 #~###### 标题。
// 其他 Markdown 标记原样保留为纯文本。解析前先经 plainNoticeText 剥离 HTML 标签/实体。

internal data class NoticeSpan(
    val text: String,
    val bold: Boolean,
    val href: String? = null
)

internal fun parseNoticeMarkdown(content: String): List<List<NoticeSpan>> {
    val text = plainNoticeText(content)
    if (text.isEmpty()) return emptyList()
    return text.split("\n").map(::parseNoticeLine)
}

private val noticeHeadingRegex = Regex("#{1,6}[ \\t]+(.*)")

private fun parseNoticeLine(line: String): List<NoticeSpan> {
    val heading = noticeHeadingRegex.matchEntire(line)
    if (heading != null) return parseNoticeSpans(heading.groupValues[1].trim(), bold = true)
    return parseNoticeSpans(line, bold = false)
}

private fun parseNoticeSpans(text: String, bold: Boolean): List<NoticeSpan> {
    val spans = mutableListOf<NoticeSpan>()
    val plain = StringBuilder()
    fun flushPlain() {
        if (plain.isNotEmpty()) {
            spans.add(NoticeSpan(plain.toString(), bold))
            plain.clear()
        }
    }
    var index = 0
    while (index < text.length) {
        if (!bold && text.startsWith("**", index)) {
            val end = text.indexOf("**", index + 2)
            if (end > index + 2) {
                flushPlain()
                spans.addAll(parseNoticeSpans(text.substring(index + 2, end), bold = true))
                index = end + 2
                continue
            }
        }
        if (text[index] == '[') {
            val link = matchNoticeLink(text, index)
            if (link != null) {
                flushPlain()
                spans.add(NoticeSpan(link.text, bold, link.href))
                index = link.end
                continue
            }
        }
        plain.append(text[index])
        index++
    }
    flushPlain()
    return spans
}

private class NoticeLink(val text: String, val href: String, val end: Int)

private fun matchNoticeLink(text: String, start: Int): NoticeLink? {
    val labelEnd = text.indexOf(']', start + 1)
    if (labelEnd < 0 || labelEnd + 1 >= text.length || text[labelEnd + 1] != '(') return null
    val hrefEnd = text.indexOf(')', labelEnd + 2)
    if (hrefEnd < 0) return null
    val label = text.substring(start + 1, labelEnd).trim()
    val href = cleanNoticeHref(text.substring(labelEnd + 2, hrefEnd))
    if (label.isEmpty() || href.isEmpty()) return null
    return NoticeLink(label, href, hrefEnd + 1)
}

private fun cleanNoticeHref(raw: String): String {
    // 面板网页主题用 #eztheme-btn 锚点标记按钮样式，客户端渲染时去掉
    val href = raw.trim().replace(Regex("#eztheme-btn$", RegexOption.IGNORE_CASE), "")
    if (href.any { it.isWhitespace() }) return ""
    if (!href.startsWith("http://", ignoreCase = true) && !href.startsWith("https://", ignoreCase = true)) return ""
    return href
}

@Composable
internal fun NoticeMarkdownText(
    paragraphs: List<List<NoticeSpan>>,
    modifier: Modifier = Modifier,
    color: Color = Color.Unspecified
) {
    val context = LocalContext.current
    val annotated = remember(paragraphs, context) { noticeAnnotatedString(paragraphs, context) }
    Text(annotated, modifier = modifier, color = color)
}

private fun noticeAnnotatedString(
    paragraphs: List<List<NoticeSpan>>,
    context: Context
): AnnotatedString = buildAnnotatedString {
    paragraphs.forEachIndexed { index, paragraph ->
        if (index > 0) append('\n')
        for (span in paragraph) appendNoticeSpan(span, context)
    }
}

private fun AnnotatedString.Builder.appendNoticeSpan(span: NoticeSpan, context: Context) {
    val href = span.href
    val style = if (span.bold) SpanStyle(fontWeight = FontWeight.Bold) else null
    if (href == null) {
        if (style == null) append(span.text) else withStyle(style) { append(span.text) }
        return
    }
    // styles 留空让 miuix Text 自动套主题主色 + 下划线的链接样式
    val link = LinkAnnotation.Url(
        url = href,
        linkInteractionListener = { BrowserOpener.open(context, href) }
    )
    withLink(link) {
        if (style == null) append(span.text) else withStyle(style) { append(span.text) }
    }
}
